### ISSUE TITLE
Will WSL2 Make ZOL Usable?

### ISSUE THREAD

**Issue by i3fe-at-work** [openzfsonwindows/ZFSin#143]
I am sorry if this is not the correct place for this question. I am wondering if WSL2 will allow Windows to use ZFS on Linux for ZFS support?

**Comment by vmlemon** [comment]
I can't speak officially for this project, since I'm only a user/tester, but I suspect that it'd be down to how easy it is to patch, and replace Microsoft's custom Linux kernel, and how much access to raw devices, their hypervisor will give you. 

There's also the problem of "What happens, if MS ever get bored with WSL (as they did with Interix/Services for UNIX), and discontinue it?" (there goes your ZFS!), and the issue of compatibility - not every machine can run the very latest builds of Windows 10 (or Hyper-V), but almost anything that can run W10, can run ZoW/ZFSin (I test it on a poverty-spec Lenovo IdeaPad, with non-upgradable components, and external storage media, so it seems to scale :)), and having a native driver opens the door to backporting it to older versions of Windows (or cross-porting it, to ReactOS, maybe?), if anyone wants to do that, for whatever reason...

Also, Windows NT has unique file system features (reparse points/junctions/streams, native ACLs, and OS/2-style extended attributes, etc.), that wouldn't necessarily work well, on a ZFS-based shared drive (which is what WSL presents, if I remember correctly; and, some applications won't even work, on a "network share"), so having a native driver gives you a higher level of fidelity, and compatibility with third-party (and, MS) applications, once all of the weird bugs, and glitches are ironed out (and, considering how much ugly, non-standard I/O Windows applications do, from what I've seen, there's plenty of them - but, once you do fix them, more stuff suddenly starts to work).

So, in short, there's room for both approaches to doing ZFS-on-Windows. Long live ZFSin! :)

**Comment by lundman** [comment]
It would be very neat if WSL2 could get ZoL to work, and it might even be tempting to lend a hand with that. Luckily there is plenty of room in "this town" and lots of cross over, so no reason for anyone to stop working :)

I want to have a go at WSL2 as soon as I can get a version - curious if there is anything I can add to ZFSin to make it work better with WSL2.

**Comment by i3fe-at-work** [comment]
Great points all.

**Comment by vmlemon** [comment]
Having thought some more, there's some other interesting ideas, and issues to explore, involving nested ZVOLs (having one hosted by one implementation, and having the child accessed under the other), having a switch to change between mounting pools under either ZFSin, or ZoL, and dealing with incompatible feature flags (What happens if you hit something like spacemap_v2, or Linux quotas, that works in one implementation, but not the other?), and sharing datasets using Windows, or Linux implementations of SMB/NFS/AFP/internal 9P.

**Comment by yurikoles** [comment]
Meanwhile, [WSL 2 is now available in Windows Insiders](https://devblogs.microsoft.com/commandline/wsl-2-is-now-available-in-windows-insiders/).

**Comment by ghfields** [comment]
There has been a WSL2 + zfs writeup.  I've not tried but.... https://wsl.dev/wsl2-kernel-zfs/
